File: sidecar/src/services/proxy_manager.py
import random
from typing import List, Optional
import httpx

class ProxyManager:
    def __init__(self, proxies: List[str]):
        self.proxies = proxies
        self.current_index = 0

    def get_proxy(self) -> Optional[str]:
        if not self.proxies:
            return None
        proxy = self.proxies[self.current_index]
        return proxy

    def rotate(self):
        if not self.proxies:
            return
        self.current_index = (self.current_index + 1) % len(self.proxies)

    def get_client_args(self) -> dict:
        proxy = self.get_proxy()
        if proxy:
            return {"proxy": proxy}
        return {}

File: sidecar/internal/scraper/proxy_manager.go
package scraper

import (
	"sync"
)

type ProxyManager struct {
	proxies      []string
	currentIndex int
	mu           sync.RWMutex
}

func NewProxyManager(proxies []string) *ProxyManager {
	return &ProxyManager{
		proxies:      proxies,
		currentIndex: 0,
	}
}

func (pm *ProxyManager) GetProxy() string {
	pm.mu.RLock()
	defer pm.mu.RUnlock()

	if len(pm.proxies) == 0 {
		return ""
	}
	return pm.proxies[pm.currentIndex]
}

func (pm *ProxyManager) Rotate() {
	pm.mu.Lock()
	defer pm.mu.Unlock()

	if len(pm.proxies) == 0 {
		return
	}
	pm.currentIndex = (pm.currentIndex + 1) % len(pm.proxies)
}

func (pm *ProxyManager) SetProxies(proxies []string) {
	pm.mu.Lock()
	defer pm.mu.Unlock()
	pm.proxies = proxies
	pm.currentIndex = 0
}
